pkg/controllers/vdb/readonlyonlineupgrade_reconciler.go

```diff
@@ -448,23 +451,85 @@ func (o *ReadOnlyOnlineUpgradeReconciler) isMatchingSubclusterType(sts *appsv1.S
 	}
 
-	if img != o.Vdb.Spec.Image {
-		scName := sts.Labels[vmeta.SubclusterNameLabel]
-		o.Log.Info("rerouting client traffic from subcluster", "name", scName)
-		if err := o.routeClientTraffic(ctx, scName, true); err != nil {
-			return ctrl.Result{}, err
+	if img == o.Vdb.Spec.Image {
+		return ctrl.Result{}, nil
+	}
+	scName := sts.Labels[vmeta.SubclusterNameLabel]
+	o.Log.Info("rerouting client traffic from subcluster", "name", scName)
+	if err := o.routeClientTraffic(ctx, scName, true); err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	o.Log.Info("starting check for active connections in subcluster", "name", scName)
+	return o.runSubclusterDrain(ctx, scName)
+}
+
+// Execute the draining for a particular subcluster
+// It will timeout after value set in active-connections-drain-seconds annotation, killing connections
+// This will re-use some code from drainnode_reconciller
+func (o *ReadOnlyOnlineUpgradeReconciler) runSubclusterDrain(ctx context.Context, scName string) (ctrl.Result, error) {
+	// The annotation we'll be using for this subcluster
+	drainSubclusterAnnotation := vmeta.GenSubclusterDrainStartAnnotationName(scName)
+
+	// Check if any OTHER subcluster drain annotation is present
+	if o.Vdb.IsOtherSubclusterDraining(scName) {
+		return ctrl.Result{}, nil
+	}
+
+	// Check for active connections
+	sessionIds, err := findSubclusterActiveSessions(ctx, o.Vdb, o.VRec, o.PFacts, scName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If no active connections found, unset annotation and return
+	if len(sessionIds) == 0 {
+		return ctrl.Result{}, removeDrainStartAnnotation(ctx, o.Vdb, o.VRec, drainSubclusterAnnotation)
+	}
+
+	timeoutInt := o.Vdb.GetActiveConnectionsDrainSeconds()
+	// If timeout is zero, return
+	if timeoutInt == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	drainStartTimeStr, found := o.Vdb.Annotations[drainSubclusterAnnotation]
+	// If drain start time annotation is not set, we set it and requeue after 1s
+	if !found {
+		o.Log.Info("Starting draining before upgrade")
+		return ctrl.Result{RequeueAfter: 1 * time.Second}, setDrainStartAnnotation(ctx, o.Vdb, o.VRec, drainSubclusterAnnotation)
+	}
+	var drainStartTime time.Time
+	drainStartTime, err = time.Parse(time.RFC3339, drainStartTimeStr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	elapsed := time.Since(drainStartTime)
+	timeout := time.Duration(timeoutInt) * time.Second
+	// If timeout has expired, kill connections and unset annotation
+	o.Log.Info("Draining in progress", "start", drainStartTime, "elapsed", elapsed, "timeout", timeout)
+	if elapsed >= timeout {
+		o.VRec.Eventf(o.Vdb, corev1.EventTypeWarning, events.DrainSubclusterTimeout,
+			"Timed out while draining connections for subcluster '%s'; killing connections", scName)
+
+		pf, found := o.PFacts.FindFirstUpPod(true, scName)
+
+		// If we cannot find a pod to use for killing, just skip
+		if found {
+			killSessions(ctx, o.VRec, o.PFacts.PRunner, sessionIds, pf)
```

Comment:
If we just skip, the annotation is still there so next time we will retry again for sc. I think if no sc nodes are up, it obviously means there will be no active connections so we should remove the annotation (if set) so next time we skip that sc.
What do you think?